### PR TITLE
[SPARK-27637][Shuffle] For nettyBlockTransferService, if IOException occurred while fetching data, check whether relative executor is alive  before retry

### PR DESCRIPTION
**Pull request by turboFei**
## What changes were proposed in this pull request?

There are several kinds of shuffle client, blockTransferService and externalShuffleClient.

For the externalShuffleClient,  there are relative external shuffle service, which guarantees  the shuffle block data and regardless the  state of executors.

For the blockTransferService, it is used to fetch broadcast block, and fetch the shuffle data when external shuffle service is not enabled. 

When fetching data by using blockTransferService, the shuffle client would connect relative executor's blockManager, so if the relative executor is dead, it would never fetch successfully.

When spark.shuffle.service.enabled is true and spark.dynamicAllocation.enabled is true,  the executor will be removed while it has been idle  for more than idleTimeout.

If a blockTransferService create connection to relative executor successfully, but the relative executor is removed when beginning to fetch broadcast block, it would retry (see RetryingBlockFetcher), which is Ineffective.

If the spark.shuffle.io.retryWait and spark.shuffle.io.maxRetries is big,  such as 30s and 10 times, it would waste 5 minutes.

In this PR, we check whether relative executor is alive before retry.
## How was this patch tested?
Unit test.

### CODE REVIEW

**Comment by turboFei** [comment]
@cloud-fan @gatorsmile

**Review by felixcheung — COMMENTED** [review]
does this account for the cases when external shuffle service is enabled?

**Comment by turboFei** [comment]
> does this account for the cases when external shuffle service is enabled?
This executorAliveChecker is only implemented in nettyBlockTransferService.
When external shuffle service is enabled,  nettyBlockTransferService is used to fetch broadcast block data and externalShuffleClient is used to fetch shuffle block data.
Therefore,  this account for fetching broadcast block when external shuffle service is enabled.
When external shuffle service is not enabled, this account for fetching both shuffle block data and broadcast data.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RetryingBlockFetcher.java [hunk 1]
```diff
@@ -178,7 +202,11 @@ private synchronized boolean shouldRetry(Throwable e) {
     boolean isIOException = e instanceof IOException
```
Comment (cloud-fan) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RetryingBlockFetcher.java]: what is the exception type when the executor is dead during shuffle fetch?

Comment (turboFei) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RetryingBlockFetcher.java]: It's IOException, just like bellow
```
2019-04-26 12:18:50,291 [26151] - WARN  [shuffle-client-4-1:TransportChannelHandler@78] - Exception in connection from hostname/ipAddress:port
java.io.IOException: Connection reset by peer
```

File: core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala [hunk 2]
```diff
@@ -117,11 +125,26 @@ private[spark] class NettyBlockTransferService(
         }
       }
 
+      val executorAliveChecker = new RetryingBlockFetcher.ExecutorAliveChecker {
+        override def check(): Boolean = {
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala]: how expensive is it?

Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala]: BTW I vaguely remember that we've already tracked the alive executors at the driver side, can you double check?

Comment (turboFei) [reply core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala]: > how expensive is it?

It would be invoked only after the IOException corrupted, so the frequency is not high.
For the driver, it only check whether a executorId is contained in a map with O(1) constant time.
For the executor, its cost is a rpc request with driver, which is not expensive .
@cloud-fan

Comment (turboFei) [reply core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala]: > BTW I vaguely remember that we've already tracked the alive executors at the driver side, can you double check?

@cloud-fan Ok, I will check it. 
 Ps:  We encountered this problem in our  production environment recently with spark-2.3.2.

Comment (turboFei) [reply core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala]: > BTW I vaguely remember that we've already tracked the alive executors at the driver side, can you double check?

@cloud-fan  Hi, I have checked the relative code of  NettyBlockTransferService  and RetryingBlockFetcher.

When the executor is dead during shuffle fetch, the retryingBlockFetcher only  just judge whether the exception is IOException and has remaining retries,  it does not  track the alive executors at the driver side.

### case 

I encountered this problem in our production environment recently with spark-2.3.2, external shuffle service and dynamicAllocation are  enabled.

When fetching broadcast data by NettyBlockTransferService, it created connection to executor_a successfully. Then executor_a was be removed because it has been idle for more then relative timeout.

So, NettyBlockTransferService catched an IOException and would retry until has no remaining retries.

This is the log of executor_a(spark.dynamicAllocation.executorIdleTimeout=40s), we can see that it was removed at 2019-04-26 12:18:49.
```
    2019-04-26 12:18:09,393 [7363] - INFO  [Executor task launch worker for task 1357:Logging$class@54] - Finished task 831.0 in stage 3.0 (TID 1357). 2046 bytes result sent to driver
    2019-04-26 12:18:09,397 [7367] - INFO  [Executor task launch worker for task 1358:Logging$class@54] - Finished task 968.0 in stage 3.0 (TID 1358). 2951 bytes result sent to driver
    2019-04-26 12:18:49,838 [47808] - ERROR [SIGTERM handler:SignalUtils$$anonfun$registerLogger$1$$anonfun$apply$1@43] - RECEIVED SIGNAL TERM
    2019-04-26 12:18:49,843 [47813] - INFO  [Thread-3:Logging$class@54] - Shutdown hook called
    2019-04-26 12:18:49,844 [47814] - INFO  [Thread-3:Logging$class@54] - Shutdown hook called
```
This is the log of NettyBlockTransferService. We can see that after executor_a is removed at 2019-04-26 12:18:49, it catched an IOException at 2019-04-26 12:18:50, and would retry.

```
    2019-04-26 12:18:49,848 [25708] - INFO  [Executor task launch worker for task 1689:Logging$class@54] - Started reading broadcast variable 5
    2019-04-26 12:18:49,906 [25766] - INFO  [Executor task launch worker for task 1689:TransportClientFactory@254] - Successfully created connection to hadoop3977.jd.163.org/10.196.64.218:38939 after 1 ms (0 ms spent in bootstraps)
    2019-04-26 12:18:50,291 [26151] - WARN  [shuffle-client-4-1:TransportChannelHandler@78] - Exception in connection from hadoop3977.jd.163.org/10.196.64.218:38939
    java.io.IOException: Connection reset by peer
    	at sun.nio.ch.FileDispatcherImpl.read0(Native Method)
    	at sun.nio.ch.SocketDispatcher.read(SocketDispatcher.java:39)
    	at sun.nio.ch.IOUtil.readIntoNativeBuffer(IOUtil.java:223)
    ......
    ......
    2019-04-26 12:18:50,296 [26156] - INFO  [shuffle-client-4-1:RetryingBlockFetcher@164] - Retrying fetch (1/30) for 1 outstanding blocks after 20000 ms
```

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RetryingBlockFetcher.java [hunk 3]
```diff
@@ -178,7 +202,11 @@ private synchronized boolean shouldRetry(Throwable e) {
...
       || (e.getCause() != null && e.getCause() instanceof IOException);
     boolean hasRemainingRetries = retryCount < maxRetries;
-    return isIOException && hasRemainingRetries;
```
[the remark this answers is not in the document]
Comment (turboFei) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RetryingBlockFetcher.java]: @cloud-fan Hi, thanks for your review~

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RetryingBlockFetcher.java [hunk 4]
```diff
@@ -61,6 +62,16 @@ void createAndStart(String[] blockIds, BlockFetchingListener listener)
          throws IOException, InterruptedException;
   }
 
+  /**
+   * Used to check whether the executor which maintain the block data to fetch is alive.
+   */
+  public interface ExecutorAliveChecker {
```
Comment (cloud-fan) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RetryingBlockFetcher.java]: do we still need it?

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RetryingBlockFetcher.java [hunk 5]
```diff
@@ -143,11 +167,22 @@ private void fetchAllOutstanding() {
       logger.error(String.format("Exception while beginning fetch of %s outstanding blocks %s",
         blockIdsToFetch.length, numRetries > 0 ? "(after " + numRetries + " retries)" : ""), e);
 
-      if (shouldRetry(e)) {
```
Comment (cloud-fan) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RetryingBlockFetcher.java]: shouldn't we just change this line to `if (shouldRetry(e) && !executorAliveChecker.check())`?

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RetryingBlockFetcher.java [hunk 6]
```diff
@@ -178,6 +213,12 @@ private synchronized boolean shouldRetry(Throwable e) {
     boolean isIOException = e instanceof IOException
       || (e.getCause() != null && e.getCause() instanceof IOException);
     boolean hasRemainingRetries = retryCount < maxRetries;
+    if (isIOException && hasRemainingRetries && executorAliveChecker != null) {
+       if (!executorAliveChecker.check()) {
+         throw new ExecutorDeadException("The relative remote executor," +
```
Comment (cloud-fan) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RetryingBlockFetcher.java]: can we throw the exception earlier? like in `fetchStarter.createAndStart`.

File: core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala [hunk 7]
```diff
@@ -325,6 +328,10 @@ class BlockManagerMasterEndpoint(
     }.toMap
   }
 
+  private def isExecutorAlive(executorId: String): Boolean = {
+    blockManagerIdByExecutor.contains(executorId)
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala]: This is just a one-line method, can we inline it?

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RetryingBlockFetcher.java (hunk 5 above)
Comment (turboFei) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RetryingBlockFetcher.java]: > shouldn't we just change this line to `if (shouldRetry(e) && !executorAliveChecker.check())`?
Done.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RetryingBlockFetcher.java (hunk 6 above)
Comment (turboFei) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RetryingBlockFetcher.java]: Yes, done.

File: core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala (hunk 7 above)
Comment (turboFei) [reply core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala]: > This is just a one-line method, can we inline it?
Done.
@cloud-fan  Thanks for your valuable advices, I think there are no other problems.

File: core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala [hunk 8]
```diff
@@ -117,11 +123,27 @@ private[spark] class NettyBlockTransferService(
         }
       }
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala]: I think this patch is overcomplicated. We just need to change here
```
val blockFetchStarter = new RetryingBlockFetcher.BlockFetchStarter {
  override def createAndStart(blockIds: Array[String], listener: BlockFetchingListener) {
    val client = clientFactory.createClient(host, port)
    try {
      new OneForOneBlockFetcher(client, appId, execId, blockIds, listener,
        transportConf, tempFileManager).start()
    } catch (IOException e) {
      if (driverEndPointRef.askSync[Boolean](IsExecutorAlive(execId))) {
        throw e
      } else {
        new ExecutorDeadException
      }
    }
  }
}
```

Comment (turboFei) [reply core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala]: Amazing, I  will modify the code, thanks.

Comment (turboFei) [reply core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala]: @cloud-fan  I have made code refactoring, with extremely admire. Thanks!

**Comment by turboFei** [comment]
thanks @cloud-fan

File: common/network-common/src/main/java/org/apache/spark/network/client/ExecutorDeadException.java [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.network.client;
+
+/**
+ * General exception caused by relative remote executor is dead when fetching block.
+ */
+public class ExecutorDeadException extends RuntimeException {
```
Comment (cloud-fan) [inline common/network-common/src/main/java/org/apache/spark/network/client/ExecutorDeadException.java]: If this exception is thrown at executor side, we should extend `SparkException`.

File: core/src/main/scala/org/apache/spark/SparkEnv.scala [hunk 10]
```diff
@@ -328,15 +328,15 @@ object SparkEnv extends Logging {
...
-    val blockTransferService =
-      new NettyBlockTransferService(conf, securityManager, bindAddress, advertiseAddress,
-        blockManagerPort, numUsableCores)
-
     val blockManagerMaster = new BlockManagerMaster(registerOrLookupEndpoint(
       BlockManagerMaster.DRIVER_ENDPOINT_NAME,
       new BlockManagerMasterEndpoint(rpcEnv, isLocal, conf, listenerBus)),
       conf, isDriver)
 
+    val blockTransferService =
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/SparkEnv.scala]: unnecessary change now?

File: core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala [hunk 11]
```diff
@@ -62,6 +66,8 @@ private[spark] class NettyBlockTransferService(
   private[this] var clientFactory: TransportClientFactory = _
   private[this] var appId: String = _
 
+  private implicit val askExecutionContext = ExecutionContext.global
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala]: is this still needed?

**Comment by cloud-fan** [comment]
Is it possible to add a unit test for it?

File: common/network-common/src/main/java/org/apache/spark/network/client/ExecutorDeadException.java (hunk 9 above)
Comment (turboFei) [reply common/network-common/src/main/java/org/apache/spark/network/client/ExecutorDeadException.java]: ok, thanks

File: core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala (hunk 11 above)
Comment (turboFei) [reply core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala]: Sorry, I will remove it.

**Comment by turboFei** [comment]
> Is it possible to add a unit test for it?

I will try my best.

**Comment by turboFei** [comment]
I have added a unit test, I create a driverEndpointRef which always return false when ask any message.
And let mock client throw IOException when sendRpc, then the nettyBlockTransferService will get if relative executors is alive, which must be false.
At last, I let the blockFetchListener throw ExecutorDeadException when it fetch failed for ExecutorDeadException.

**Comment by turboFei** [comment]
@cloud-fan I also add relative unit test to test retryingBlockFetcher. Which assert that retryingBlockFetcher does not retry because the relative executor has been dead.

File: core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala [hunk 12]
```diff
@@ -112,8 +117,18 @@ private[spark] class NettyBlockTransferService(
...
         override def createAndStart(blockIds: Array[String], listener: BlockFetchingListener) {
           val client = clientFactory.createClient(host, port)
-          new OneForOneBlockFetcher(client, appId, execId, blockIds, listener,
-            transportConf, tempFileManager).start()
+          try {
+            new OneForOneBlockFetcher(client, appId, execId, blockIds, listener,
+              transportConf, tempFileManager).start()
+          } catch {
+            case e: IOException =>
+              if (driverEndPointRef.askSync[Boolean](IsExecutorAlive(execId))) {
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala]: nit: add a safeguard when `driverEndPointRef` is null.

File: core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala [hunk 13]
```diff
@@ -49,7 +53,8 @@ private[spark] class NettyBlockTransferService(
     bindAddress: String,
     override val hostName: String,
     _port: Int,
-    numCores: Int)
+    numCores: Int,
+    driverEndPointRef: RpcEndpointRef = null)
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala]: when will we use the default null value? only in test?

File: core/src/test/scala/org/apache/spark/network/netty/NettyBlockTransferServiceSuite.scala [hunk 14]
```diff
@@ -77,6 +87,68 @@ class NettyBlockTransferServiceSuite
...
+      Array("block1"), listener, mock(classOf[DownloadFileManager])))
+
+    val blockFetchStarter = new RetryingBlockFetcher.BlockFetchStarter {
+      override def createAndStart(blockIds: Array[String], listener: BlockFetchingListener) {
+        try {
+          new OneForOneBlockFetcher(client, "appId", "execId", blockIds, listener,
+            mock(classOf[TransportConf]), mock(classOf[DownloadFileManager])).start()
+        } catch {
+          case e: IOException =>
+            if (driverEndpointRef.askSync[Boolean](IsExecutorAlive("execId"))) {
```
Comment (cloud-fan) [inline core/src/test/scala/org/apache/spark/network/netty/NettyBlockTransferServiceSuite.scala]: This is the code we want to test, we should not put it in the test...

Comment (cloud-fan) [reply core/src/test/scala/org/apache/spark/network/netty/NettyBlockTransferServiceSuite.scala]: one principal to write a regression test: it should fail without the patch

**Comment by turboFei** [comment]
> when will we use the default null value? only in test?
Yes, the default null value is only used in test.
So, I think we don't need add safeguard when driverEndPointRef is null.

> This is the code we want to test, we should not put it in the test...

Get it, I wanted to prove that the retryingBlockFetch would not retry when the relative executor is dead. I will remove this part. 
@cloud-fan Thanks

File: core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala (hunk 13 above)
Comment (turboFei) [reply core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala]: yes.

File: core/src/test/scala/org/apache/spark/network/netty/NettyBlockTransferServiceSuite.scala (hunk 14 above)
Comment (turboFei) [reply core/src/test/scala/org/apache/spark/network/netty/NettyBlockTransferServiceSuite.scala]: Get it, I wanted to prove that the retryingBlockFetch would not retry when the relative executor is dead. I have remove this part.

File: core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala (hunk 12 above)
Comment (turboFei) [reply core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala]: > nit: add a safeguard when `driverEndPointRef` is null.
we use the default null value only in test. So, I think we may not need add safeguard when `driverEndPointRef` is null.

@cloud-fan

Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala]: then let's add an assert

File: core/src/test/scala/org/apache/spark/network/netty/NettyBlockTransferServiceSuite.scala [hunk 15]
```diff
@@ -77,6 +85,44 @@ class NettyBlockTransferServiceSuite
...
+
+    service0 = createService(port, driverEndpointRef)
+    val clientFactoryField = service0.getClass.getField(
+      "org$apache$spark$network$netty$NettyBlockTransferService$$clientFactory")
+    clientFactoryField.setAccessible(true)
+    clientFactoryField.set(service0, clientFactory)
+
+    // The ExecutorDeadException is thrown by listener when fetch failed for ExecutorDeadException.
+    intercept[ExecutorDeadException](service0.fetchBlocks("localhost", port, "exec1",
+      Array("block1"), listener, mock(classOf[DownloadFileManager])))
```
Comment (cloud-fan) [inline core/src/test/scala/org/apache/spark/network/netty/NettyBlockTransferServiceSuite.scala]: We need to make sure that we only send one fetch request. We can count the calls of `TransportClientFactory.createClint`, the mock lib can do it.

File: core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala (hunk 12 above)
Comment (turboFei) [reply core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala]: > then let's add an assert

done.

@cloud-fan

File: core/src/test/scala/org/apache/spark/network/netty/NettyBlockTransferServiceSuite.scala (hunk 15 above)
Comment (turboFei) [reply core/src/test/scala/org/apache/spark/network/netty/NettyBlockTransferServiceSuite.scala]: > We need to make sure that we only send one fetch request. We can count the calls of `TransportClientFactory.createClint`, the mock lib can do it.

I will try.

Comment (turboFei) [reply core/src/test/scala/org/apache/spark/network/netty/NettyBlockTransferServiceSuite.scala]: > We need to make sure that we only send one fetch request. We can count the calls of `TransportClientFactory.createClint`, the mock lib can do it.

done. thanks, this is my first time to use the `verify`. @cloud-fan

**Comment by turboFei** [comment]
shall we let sparkQA test this patch?  I am not a contributor currently.
@cloud-fan

File: core/src/test/scala/org/apache/spark/network/netty/NettyBlockTransferServiceSuite.scala (hunk 15 above)
Comment (turboFei) [reply core/src/test/scala/org/apache/spark/network/netty/NettyBlockTransferServiceSuite.scala]: gentle ping and thanks for your review. @cloud-fan

**Comment by cloud-fan** [comment]
ok to test

**Comment by turboFei** [comment]
Test Passed. Thanks for you review, I have gained a lot. Thanks very much. @cloud-fan

File: core/src/test/scala/org/apache/spark/network/netty/NettyBlockTransferServiceSuite.scala [hunk 16]
```diff
@@ -77,6 +85,51 @@ class NettyBlockTransferServiceSuite
...
+      "org$apache$spark$network$netty$NettyBlockTransferService$$clientFactory")
+    clientFactoryField.setAccessible(true)
+    val originClientFactory = clientFactoryField.get(service0)
+    clientFactoryField.set(service0, clientFactory)
+
+    // The ExecutorDeadException is thrown by listener when fetch failed for ExecutorDeadException.
+    intercept[ExecutorDeadException](service0.fetchBlocks("localhost", port, "exec1",
+      Array("block1"), listener, mock(classOf[DownloadFileManager])))
+    assert(createClientCount === 1)
+    clientFactoryField.set(service0, originClientFactory)
```
Comment (cloud-fan) [inline core/src/test/scala/org/apache/spark/network/netty/NettyBlockTransferServiceSuite.scala]: This is a unit test, we don't need to set the original factory back.

File: core/src/test/scala/org/apache/spark/network/netty/NettyBlockTransferServiceSuite.scala [hunk 17]
```diff
@@ -77,6 +85,51 @@ class NettyBlockTransferServiceSuite
...
+    when(clientFactory.createClient(any(), any())).thenAnswer(_ => {
+      createClientCount += 1
+      client
+    })
+
+    val listener = mock(classOf[BlockFetchingListener])
+    // This is used to throw a ExecutorDeadException for unit test, when the fetch is failed
+    // because of ExecutorDeadException.
+    when(listener.onBlockFetchFailure(any(), any(classOf[ExecutorDeadException])))
+      .thenAnswer(_ => {throw new ExecutorDeadException("Executor is dead.")})
```
Comment (cloud-fan) [inline core/src/test/scala/org/apache/spark/network/netty/NettyBlockTransferServiceSuite.scala]: nit:
```
var hitExecutorDeadException = false
when(listener.onBlockFetchFailure(any(), any(classOf[ExecutorDeadException])))
  .thenAnswer(_ => {hitExecutorDeadException = true})
...
service0.fetchBlocks...
assert(createClientCount === 1)
assert(hitExecutorDeadException)
```

**Comment by cloud-fan** [comment]
LGTM, also cc @jiangxb1987 @zsxwing

File: core/src/test/scala/org/apache/spark/network/netty/NettyBlockTransferServiceSuite.scala (hunk 16 above)
Comment (turboFei) [reply core/src/test/scala/org/apache/spark/network/netty/NettyBlockTransferServiceSuite.scala]: done.

File: core/src/test/scala/org/apache/spark/network/netty/NettyBlockTransferServiceSuite.scala (hunk 17 above)
Comment (turboFei) [reply core/src/test/scala/org/apache/spark/network/netty/NettyBlockTransferServiceSuite.scala]: done.

File: core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala [hunk 18]
```diff
@@ -112,8 +117,19 @@ private[spark] class NettyBlockTransferService(
...
           val client = clientFactory.createClient(host, port)
-          new OneForOneBlockFetcher(client, appId, execId, blockIds, listener,
-            transportConf, tempFileManager).start()
+          try {
+            new OneForOneBlockFetcher(client, appId, execId, blockIds, listener,
+              transportConf, tempFileManager).start()
+          } catch {
+            case e: IOException =>
+              assert(driverEndPointRef != null)
+              if (driverEndPointRef.askSync[Boolean](IsExecutorAlive(execId))) {
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala]: If `askSync` fails, we should throw the original exception.

Comment (turboFei) [reply core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala]: > If `askSync` fails, we should throw the original exception.
Thanks,  I used to think that if askSync fails, it would also throw another IOException, whose stackTrace is similar with origin IOException.
Done.

Comment (turboFei) [reply core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala]: should we remove the `assert(driverEndPointRef != null)`, because the invocation of driverEndPointRef is in `Try`. @cloud-fan

**Comment by turboFei** [comment]
This test(https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105561/console) has been hanged for 2 hours.
So I force push a commit to trigger a new test.

**Comment by turboFei** [comment]
> LGTM, also cc @jiangxb1987 @zsxwing

gentle ping @jiangxb1987  @zsxwing

**Comment by cloud-fan** [comment]
thanks, merging to master!
